Fix flaky test_setting_concurrent_rebuild_limit Extend Volume Rebuild Timeout

roger-ryao opened the pull request:
https://github.com/longhorn/longhorn/issues/6382

The [Volume Replica Rebuilding Speedup](https://github.com/longhorn/longhorn/issues/4783) feature was introduced as an enhancement in `1.4.0`

The current timeout value for checking the rebuild status may not be sufficient for `v1.3.x`.

To address this, we propose extending the timeout for checking the rebuild volume status on `v1.3.x` branch.

The execution time for each test case to approximately 15-20 minutes.


![Screenshot_20230802_172123](https://github.com/longhorn/longhorn-tests/assets/104337648/6643a6bf-086d-4802-ae96-4f6816193f3e)


Private image test result 
- [x] coretest : https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4727/
- [x] test case test : https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4725/

yangchiu commented:
According to the description https://github.com/longhorn/longhorn/issues/6382

```python
        wait_for_rebuild_complete(client, volume2_name)
        wait_for_rebuild_complete(client, volume1_name)
    
        # Step 2-1
        # Step 2-2
>       wait_for_volume_healthy(client, volume1_name)
```

If `wait_for_rebuild_complete` already passes, doesn't it mean the rebuilding has completed? Why does it still have rebuild timeout issue here?

roger-ryao commented:
> According to the description [longhorn/longhorn#6382](https://github.com/longhorn/longhorn/issues/6382)
> 
> ```python
>         wait_for_rebuild_complete(client, volume2_name)
>         wait_for_rebuild_complete(client, volume1_name)
>     
>         # Step 2-1
>         # Step 2-2
> >       wait_for_volume_healthy(client, volume1_name)
> ```
> 
> If `wait_for_rebuild_complete` already passes, doesn't it mean the rebuilding has completed? Why does it still have rebuild timeout issue here?

Hi @yangchiu :
      Yes, you are right. When I execute the test case and encounter a failure, the replica is still rebuilding. I added a rebuild timeout for that scenario, but I didn't investigate why "wait_for_rebuild_complete(client, volume1_name)" passed without any error. It seems that the function didn't correctly detect the rebuilding status in some cases, which is why I added the timeout as a temporary solution.

yangchiu commented:
>Yes, you are right. When I execute the test case and encounter a failure, the replica is still rebuilding. I added a rebuild timeout for that scenario, but I didn't investigate why "wait_for_rebuild_complete(client, volume1_name)" passed without any error. It seems that the function didn't correctly detect the rebuilding status in some cases, which is why I added the timeout as a temporary solution.

Can you help to do more experiments to investigate why? Just insert some logs and you're able to know the whole status at that moment.

roger-ryao commented:
> Can you help to do more experiments to investigate why? Just insert some logs and you're able to know the whole status at that moment.

I will take some time to investigate why the replica is still rebuilding even after wait_for_rebuild_complete(client, volume1_name) has passed without any error.

roger-ryao commented:
> > Can you help to do more experiments to investigate why? Just insert some logs and you're able to know the whole status at that moment.
> 
> I will take some time to investigate why the replica is still rebuilding even after wait_for_rebuild_complete(client, volume1_name) has passed without any error.

Hi @yangchiu 
The existing code at https://github.com/longhorn/longhorn-tests/blob/master/manager/integration/tests/test_settings.py#L856-L878 does not guarantee that the volume has actually started rebuilding. To ensure more rigor, I have added a check to verify that the rebuild status must be greater than 50% before considering `rebuild_started` as True. Before calling wait_for_rebuild_complete, I think it's essential to ensure that wait_for_rebuild_start is executed. This way, if any issues arise during wait_for_rebuild_start, we can better understand the root cause of the problem.

yangchiu commented:
> Hi @yangchiu The existing code at https://github.com/longhorn/longhorn-tests/blob/master/manager/integration/tests/test_settings.py#L856-L878 does not guarantee that the volume has actually started rebuilding. To ensure more rigor, I have added a check to verify that the rebuild status must be greater than 50% before considering `rebuild_started` as True. Before calling wait_for_rebuild_complete, I think it's essential to ensure that wait_for_rebuild_start is executed. This way, if any issues arise during wait_for_rebuild_start, we can better understand the root cause of the problem.

Why doesn't the original `assert volume1.rebuildStatus[0].state == "in_progress"` ensure the volume has started rebuilding?
Isn't the same checking used in `wait_for_rebuild_start` later also being used by you to ensure the same thing?

yangchiu commented on manager/integration/tests/test_settings.py:
```diff
@@ -896,25 +896,28 @@ def test_setting_concurrent_rebuild_limit(client, core_api, volume_name):  # NOQ
+                rebuild_started = volume1.rebuildStatus[0].progress >= 50
+
+        # Check volume2 if volume1 is not rebuilding
+        if not volume1.rebuildStatus:
```
Will a volume doesn't have `rebuildStatus`?

roger-ryao replied:
Yes, according to [the distribution scenario mentioned](https://github.com/longhorn/longhorn-tests/blob/v1.3.x/manager/integration/tests/test_settings.py#L783-L786), one of the volumes does not have a `rebuildStatus`.

yangchiu commented on manager/integration/tests/test_settings.py:
```diff
@@ -896,25 +896,28 @@ def test_setting_concurrent_rebuild_limit(client, core_api, volume_name):  # NOQ
-    wait_for_rebuild_complete(client, volume1_name)
+    assert rebuild_started
+    wait_for_rebuild_complete(client, volume2_name, 300)
+    wait_for_rebuild_start(client, volume1_name)
```
This implies `volume2` started rebuilding first, and then `volume1`.

If `volume1` started rebuilding first, this wait will fail, right?

roger-ryao replied:
Yes, you are correct. To enhance and test this situation, I will modify the wait condition to check for the status of both `volume1` and `volume2`.

roger-ryao commented:
> Why doesn't the original `assert volume1.rebuildStatus[0].state == "in_progress"` ensure the volume has started rebuilding? Isn't the same checking used in `wait_for_rebuild_start` later also being used by you to ensure the same thing?


Initially, the rebuild process starts for `Volume1` and progresses up to `13%`. However, after a certain period of time, the rebuild process for `volume1` is canceled. Subsequently, the rebuild process for `volume2` begins. This is why I added a check status process for both volumes, ensuring the rebuild progresses beyond 50%, and set the `rebuild_start` flag to `True`.

yangchiu commented:
> Initially, the rebuild process starts for `Volume1` and progresses up to `13%`. However, after a certain period of time, the rebuild process for `volume1` is canceled. Subsequently, the rebuild process for `volume2` begins. This is why I added a check status process for both volumes, ensuring the rebuild progresses beyond 50%, and set the `rebuild_start` flag to `True`.

When `volume1` rebuilding is canceled, what's the `rebuildStatus` of `volume1`? Will it return to an empty array `[]`? Or will any message present in the volume status?

roger-ryao commented:
> > Initially, the rebuild process starts for `Volume1` and progresses up to `13%`. However, after a certain period of time, the rebuild process for `volume1` is canceled. Subsequently, the rebuild process for `volume2` begins. This is why I added a check status process for both volumes, ensuring the rebuild progresses beyond 50%, and set the `rebuild_start` flag to `True`.
> 
> When `volume1` rebuilding is canceled, what's the `rebuildStatus` of `volume1`? Will it return to an empty array `[]`? Or will any message present in the volume status?

When the rebuilding of `volume1` is canceled, it will return to an empty array [].

![Screenshot_20230729_223914](https://github.com/longhorn/longhorn-tests/assets/104337648/f02d0f6f-3fad-4ede-bce2-094168feed51)

roger-ryao commented:
> > When the rebuilding of `volume1` is canceled, it will return to an empty array [].
> 
> Only the rebuilding of `volume1` will be canceled, right?
> 
> Will it happen on `volume2`?

Yes, you are correct. Based on the scenario mentioned in https://github.com/longhorn/longhorn-tests/blob/v1.3.x/manager/integration/tests/test_settings.py#L783-L786, we observed that the rebuilding process of `volume1` was canceled. However, after a certain period of time, `volume2` started its rebuilding process, and unlike `volume1`, its rebuilding process was not canceled.

yangchiu commented:
> > > When the rebuilding of `volume1` is canceled, it will return to an empty array [].
> > 
> > 
> > Only the rebuilding of `volume1` will be canceled, right?
> > Will it happen on `volume2`?
> 
> Yes, you are correct. Based on the scenario mentioned in https://github.com/longhorn/longhorn-tests/blob/v1.3.x/manager/integration/tests/test_settings.py#L783-L786, we observed that the rebuilding process of `volume1` was canceled. However, after a certain period of time, `volume2` started its rebuilding process, and unlike `volume1`, its rebuilding process was not canceled.

So does this test case fail only because it doesn't make sure `volume1`'s rebuilding is canceled, so in the `for loop`, it incorrectly thought rebuilding has been done on both of 2 volumes, and then causes the `wait_for_volume_healthy` for `volume1` failed?

roger-ryao commented:
> So does this test case fail only because it doesn't make sure `volume1`'s rebuilding is canceled, so in the `for loop`, it incorrectly thought rebuilding has been done on both of 2 volumes, and then causes the `wait_for_volume_healthy` for `volume1` failed?

You are correct in your analysis. The [Volume Replica Rebuilding Speedup feature introduced](https://github.com/longhorn/longhorn/issues/4783) in version `1.4.0`, which is why the test case doesn't fail in versions `1.4`, `1.5`, and the `master` branch. it's still possible to enhance and test the case on the master branch for further validation and improvement.

yangchiu commented:
> > So does this test case fail only because it doesn't make sure `volume1`'s rebuilding is canceled, so in the `for loop`, it incorrectly thought rebuilding has been done on both of 2 volumes, and then causes the `wait_for_volume_healthy` for `volume1` failed?
> 
> You are correct in your analysis. The [Volume Replica Rebuilding Speedup feature introduced](https://github.com/longhorn/longhorn/issues/4783) in version `1.4.0`, which is why the test case doesn't fail in versions `1.4`, `1.5`, and the `master` branch. it's still possible to enhance and test the case on the master branch for further validation and improvement.

Can't make sure `volume`'s rebuilding is canceled before enter the `for loop` solve this problem?

roger-ryao commented:
> Can't make sure `volume`'s rebuilding is canceled before enter the `for loop` solve this problem?

After discussing with @yangchiu, I will try to enhance the solution by ensuring that the volume's rebuilding is canceled before entering the for loop.

roger-ryao commented:
> > Can't make sure `volume`'s rebuilding is canceled before enter the `for loop` solve this problem?
> 
> After discussing with @yangchiu, I will try to enhance the solution by ensuring that the volume's rebuilding is canceled before entering the for loop.

Hi @yangchiu :
     I observed that when we retrieve the rebuilding status of `volume1`, the replica count remains two. Therefore, in addition to checking `volume.rebuildStatus[0].state`, I suggest confirming that `volume replicas count` is [also equal to three](https://github.com/longhorn/longhorn-tests/pull/1477/files#diff-8416f1cda57e975c3c1d0238a26790d74d71dbb3ae4b6e3cb468209d56bf1a12R904-R917). This precaution will prevent acquiring incorrect information about `volume.rebuildStatus` within the for loop.

yangchiu commented:
> 
> Hi @yangchiu : I observed that when we retrieve the rebuilding status of `volume1`, the replica count remains two. Therefore, in addition to checking `volume.rebuildStatus[0].state`, I suggest confirming that `volume replicas count` is [also equal to three](https://github.com/longhorn/longhorn-tests/pull/1477/files#diff-8416f1cda57e975c3c1d0238a26790d74d71dbb3ae4b6e3cb468209d56bf1a12R904-R917). This precaution will prevent acquiring incorrect information about `volume.rebuildStatus` within the for loop.

`replica count = 2` is because we remove one of replicas in previous step. Once the rebuilding started, the 3rd replica will be created automatically. Why we need to explicitly check this? And why does it result in _incorrect information_?

roger-ryao commented:
> > Hi @yangchiu : I observed that when we retrieve the rebuilding status of `volume1`, the replica count remains two. Therefore, in addition to checking `volume.rebuildStatus[0].state`, I suggest confirming that `volume replicas count` is [also equal to three](https://github.com/longhorn/longhorn-tests/pull/1477/files#diff-8416f1cda57e975c3c1d0238a26790d74d71dbb3ae4b6e3cb468209d56bf1a12R904-R917). This precaution will prevent acquiring incorrect information about `volume.rebuildStatus` within the for loop.
> 
> `replica count = 2` is because we remove one of replicas in previous step. Once the rebuilding started, the 3rd replica will be created automatically. Why do we need to explicitly check this? And why does it result in _incorrect information_?

Because I discovered that when we observe the rebuild process for `volume1` being canceled, the replica count for `volume1` remains at `2`. Our expectation is that when we retrieve volume information during an ongoing rebuild process, the `replica count` for the volume should be `3`.

yangchiu commented:
@roger-ryao What's the reproducibility of the original issue https://github.com/longhorn/longhorn/issues/6382 in longhorn `v1.3.x-head` tested with not modified longhorn-tests `v1.3.x-head`?

roger-ryao commented:
> @roger-ryao What's the reproducibility of the original issue [longhorn/longhorn#6382](https://github.com/longhorn/longhorn/issues/6382) in longhorn `v1.3.x-head` tested with not modified longhorn-tests `v1.3.x-head`?

Hi @yangchiu :
     The reproduction rate was 80%.

roger-ryao commented:
> @roger-ryao What's the reproducibility of the original issue [longhorn/longhorn#6382](https://github.com/longhorn/longhorn/issues/6382) in longhorn `v1.3.x-head` tested with not modified longhorn-tests `v1.3.x-head`?

Hi @yangchiu 
I attempted to rerun the test case 20 times, and it only passed once. You can find the results here:
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4751/
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4752/

When I executed the test case to address the original issue longhorn/longhorn#6382, I also discovered that we have three additional steps that are prone to failure due to timeouts during the wait for replica rebuild completion. This is why you can see that I extended the timeout for some of the steps.

yangchiu commented:
@roger-ryao Is it possible to have the test result of the latest version of this PR?

roger-ryao commented:
> @roger-ryao Is it possible to have the test result of the latest version of this PR?

Hi @yangchiu:

I have tested the private image before submitting the code, and the test results are as follows:
Test case test: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4725/testReport/tests/test_settings/

innobead commented on manager/integration/tests/test_settings.py:
```diff
@@ -891,28 +891,30 @@ def test_setting_concurrent_rebuild_limit(client, core_api, volume_name):  # NOQ
-        time.sleep(RETRY_INTERVAL)
+        time.sleep(RETRY_INTERVAL_LONG)
 
+    assert v2_rebuild_started
+    assert v1_rebuild_started
```
I am a little confused. What we want to test is volume 2 rebuilding wil run first, then after volume 1. However, here, we just check both are started to rebuild. It seems not what we want to test?

roger-ryao replied:
Hi @innobead :
     After discussing with James previously, there were concerns about the order of rebuilding [between volume 1 and volume 2](https://github.com/longhorn/longhorn-tests/blob/c9e603cf6815030305d8a0e8e504251b9431243d/manager/integration/tests/test_settings.py#L785C1-L786C62.). 
     
It was noted that there are two possible scenarios:
Volume 2's rebuilding starts first, and volume 1's rebuilding is delayed until volume 2 becomes healthy.
Volume 1's rebuilding starts first, and volume 2's rebuilding waits until volume 1 becomes healthy.

The exact order of rebuilding initiation may vary, our enhancement ensures that both volumes have been successfully under rebuilding.

innobead replied:
Then, this verification does not achieve what we want to check. Probably you can juts check there is only one rebuilding, then then all rebuilt successfully eventually. This can make the test meaningful and complete. WDYT?

yangchiu replied:
@roger-ryao Maybe you can try to print out all the rebuild status throughout the whole test case process, so we can have a clear picture of the rebuilding orders, which rebuilding is canceled/restarted, or any behavior that is unexpected.

innobead reviewed (changes requested):
Just one feedback.
